<center>

### COSC2753 - Machine Learning

# **Exploratory Data Analysis (EDA)**

<center>────────────────────────────</center>
&nbsp;

## Table of Contents

</center>

<div style="text-align: left;">

<div style="display: inline-block; text-align: left;">

<h3>

[**I. Introduction**](#introduction)

</h3>

<h3>
    
[**II. Libraries Import and Data Preprocessing**](#libraries-import-and-data-preprocessing)

</h3>

<h4>

- [1. Importing Libraries](#1-importing-libraries)

</h4>

<h4>

- [2. Global Configuration](#2-global-configuration)

</h4>

<h4>

- [3. Data Loading](#3-data-loading)

</h4>

<h3>

[**III. Data Collection**](#data-collection)

</h3>

<h3>

[**IV. Data Cleaning**](#data-cleaning)

</h3>

<h3>

[**V. Exploratory Data Analysis**](#exploratory-data-analysis)

</h3>

<h3>
    
[**VI. Conclusion**](#conclusion)

</h3>

<h3>
    
[**VII. References**](#references)

</div>

<center><em><sub>─────── End Of Section ───────</sub></em></center>

## Introduction

This notebook focuses on Exploratory Data Analysis (EDA) as the foundation for our machine learning workflow with the provided datasets. Through EDA, we will gain a comprehensive understanding of the data, including:

- Data cleaning and preprocessing.
- Features/Target variables distribution and relationships analysis.
- Correlation analysis.
- Outliers detection.
- Model selection and evaluation.

Upon finishing, this notebook will provide valuable insights of the data for subsequent stages.

<center><em><sub>─────── End Of Section ───────</sub></em></center>

## Libraries Import and Data Preprocessing 

### 1. Importing Libraries

In [516]:
# Import installed python packages
import pandas as pd  #  Data manipulation
import numpy as np  # Numerical operations
import seaborn as sns  # Visualization
import matplotlib.pyplot as plt  # Visualization
from sklearn.preprocessing import MinMaxScaler, StandardScaler  # Feature scaling
import time  # Time operations

# Import user-defined python scripts
import sys
import importlib

sys.path.append("../../")  # Root directory
modules_to_reload = [
    "scripts.outlier_detector",
    "scripts.styler",
    "scripts.neko",
]

# Reload modules if they have been modified
start_time = time.time()  # Start time
missing_modules = []  # Missing modules

# Reload modules
for module_name in modules_to_reload:
    if module_name in sys.modules:
        importlib.reload(sys.modules[module_name])
    else:
        missing_modules.append(module_name)

# Recache missing modules
if missing_modules:
    print(f"Modules {missing_modules} not found. \nRecaching...")


from scripts.outlier_detector import OutlierDetector  # Outlier detection
from scripts.styler import Styler  # Visualization
from scripts.neko import Neko  # Panda extension

end_time = time.time()  # End time

runtime = end_time - start_time  # Runtime

print("Done in:", round(runtime, 3), "seconds")

Done in: 0.002 seconds


### 2. Global Configuration

In [517]:
# Set global parameters
styler = Styler()  # Text Styler
neko = Neko()  # Panda extension
bullet = ">>>"  # Bullet point
pd.set_option("display.max_columns", None)  # Display all columns

styler.draw_box("Project has been configured successfully")  # Draw title

┌────────────────────────────────────────────┐
│  Project has been configured successfully  │
└────────────────────────────────────────────┘


### 3. Data Loading

In [518]:
# Load the datasets
try:
    df_train = pd.read_csv("../../data/raw/data_train.csv")  # Training dataset
    styler.draw_box("Datasets loaded successfully")

    print(" ")

    # Display the overview of the training dataset
    text = "Training dataset overview:"
    styled_text = styler.style(text, bold=True, italic=True, underline=True)
    print(f"{bullet} {styled_text}")
    print(neko.essense(df_train))

except Exception as e:
    styler.draw_box("Error loading datasets : " + str(e))

┌────────────────────────────────┐
│  Datasets loaded successfully  │
└────────────────────────────────┘
 
>>> Training dataset overview:
 - Number of rows: 202944
 - Number of columns: 25
 - Data types: [dtype('int64')]
 - # Missing values: 0 at columns: []
 - # Duplicate rows: 0
 - Data size: 38.7 MB


**Initial Data Exploration**

This initial exploration of the data yielded the following key observations:

- **Data Loading:** The dataset comprises approximately `200,000` rows and `25` columns. Given its size, careful selection of data cleaning and preprocessing techniques will be essential to ensure data quality and integrity for further analysis.

- **Data Type Consistency:** A positive aspect is that all columns possess the same data type (`int64`), potentially simplifying the data cleaning process.

- **Missing Value Assessment:** An initial check for missing values (`null`, `NaN`, or `empty values`) was conducted. This ensures the data is not empty or containing invalid values.

- **Duplicate Detection:** No duplicates were identified within the dataset at this stage (with the presence of the `Id` column).

&nbsp;

**Feature Explanation**

| Variable             | Description                                                                                                                                                                  | Category    |
| -------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| ID                   | Patient ID                                                                                                                                                                   | Numerical   |
| Status               | 0 = no diabetes, 1 = prediabetes or diabetes                                                                                                                                 | Categorical |
| HighBP               | 0 = no high BP, 1 = high BP                                                                                                                                                  | Categorical |
| HighChol             | 0 = no high cholesterol, 1 = high cholesterol                                                                                                                                | Categorical |
| CholCheck            | 0 = no cholesterol check in 5 years, 1 = yes cholesterol check in 5 years                                                                                                    | Categorical |
| BMI                  | Body Mass Index                                                                                                                                                              | Numerical   |
| Smoker               | 0 = no, 1 = yes (smoked at least 100 cigarettes in lifetime)                                                                                                                 | Categorical |
| Stroke               | 0 = no, 1 = yes (history of stroke)                                                                                                                                          | Categorical |
| HeartDiseaseorAttack | 0 = no, 1 = yes (coronary heart disease or myocardial infarction)                                                                                                            | Categorical |
| PhysActivity         | 0 = no, 1 = yes (physical activity in past 30 days, excluding job)                                                                                                           | Categorical |
| Fruits               | 0 = no, 1 = yes (consume fruit ≥ 1 time per day)                                                                                                                             | Categorical |
| Veggies              | 0 = no, 1 = yes (consume vegetables ≥ 1 time per day)                                                                                                                        | Categorical |
| HvyAlcoholConsump    | 0 = no, 1 = yes (heavy alcohol consumption)                                                                                                                                  | Categorical |
| AnyHealthcare        | 0 = no, 1 = yes (health care coverage)                                                                                                                                       | Categorical |
| NoDocbcCost          | 0 = no, 1 = yes (unable to see doctor due to cost)                                                                                                                           | Categorical |
| GenHlth              | Scale 1-5: 1 = excellent, 2 = very good, 3 = good, 4 = fair, 5 = poor                                                                                                        | Categorical |
| MentHlth             | Number of days in past 30 days with poor mental health (scale 1-30 days)                                                                                                     | Numerical   |
| PhysHlth             | Number of days in past 30 days with poor physical health (scale 1-30 days)                                                                                                   | Numerical   |
| DiffWalk             | 0 = no, 1 = yes (serious difficulty walking or climbing stairs)                                                                                                              | Categorical |
| Sex                  | 0 = female, 1 = male                                                                                                                                                         | Categorical |
| Age                  | 13-level age category: 1 = 18-24, 9 = 60-64, 13 = 80 or older                                                                                                                | Categorical |
| Education            | Scale 1-6: 1 = Never attended school or only kindergarten, 2 = Grades 1 through 8, 3 = Grades 9 through 11, 4 = Grade 12 or GED, 5 = College 1-3 years, 6 = College 4+ years | Categorical |
| Income               | Scale 1-8: 1 = less than $10,000, 5 = less than $35,000, 8 = $75,000 or more                                                                                                 | Categorical |
| ExtraMedTest         | Result of an extra medical test (range -100 to 100)                                                                                                                          | Numerical   |
| ExtraAlcoholTest     | Result of an extra alcohol test (range -100 to 100)                                                                                                                          | Numerical   |

**Note**:

- The 'Id' column will be excluded in accordance with project specifications.
- This initial exploration reveals that the majority of the columns contain `categorical` features. This suggests a potential focus on `categorical data analysis` techniques for further exploration and model selection.
- An initial examination of the target variable (`Status`) reveals it consists of binary values (0 or 1). This characteristic indicates a `classification` problem, where we aim to predict the category (0 or 1) for new data points. Given this binary nature, `Logistic Regression` emerges as a strong candidate for the initial model selection.


### 4. Data Preprocessing

#### 4.1. Prerequisites

In [519]:
# Remove the Id column from the dataset as required
df_train.drop("Id", axis=1, inplace=True)

styler.draw_box("Id column removed successfully")

┌──────────────────────────────────┐
│  Id column removed successfully  │
└──────────────────────────────────┘


#### 4.2. Data Overview

In [520]:
styler.draw_box("Checking for duplicates in the dataset (After removing Id column)")

# Check for duplicates in the dataset
duplicates = df_train.duplicated().sum()  # Duplicates
if duplicates > 0:
    text = f"Found {duplicates} duplicates in the dataset"
    styled_text = styler.style(text, bold=True, italic=True)
    print(f"{bullet} {styled_text}")

    # Remove duplicates
    print(f"{bullet} Removing duplicates...")

    start_time = time.time()  # Start time

    df_train.drop_duplicates(inplace=True)  # Remove duplicates

    end_time = time.time()  # End time

    runtime = end_time - start_time  # Runtime

    print(f"{bullet} Done in:", round(runtime, 3), "seconds")
else:
    text = f"No duplicates in the dataset"
    styled_text = styler.style(text, bold=True, italic=True)
    print(f"{bullet} {styled_text}")

┌─────────────────────────────────────────────────────────────────────┐
│  Checking for duplicates in the dataset (After removing Id column)  │
└─────────────────────────────────────────────────────────────────────┘
>>> Found 208 duplicates in the dataset
>>> Removing duplicates...
>>> Done in: 0.059 seconds


**Note:** 
- Following the removal of the "Id" column, an initial analysis identified 208 duplicate data points within the dataset. 

  => To ensure data integrity and quality, these duplicates will be removed from the dataset before proceeding with further analysis.


In [521]:
# Display the overview of the training dataset
styler.draw_box("Training dataset description")
df_train.describe()

┌────────────────────────────────┐
│  Training dataset description  │
└────────────────────────────────┘


HighBP       HighChol      CholCheck            BMI  \
count  202736.000000  202736.000000  202736.000000  202736.000000   
mean        0.429060       0.424631       0.962621      28.383844   
std         0.494943       0.494288       0.189688       6.614264   
min         0.000000       0.000000       0.000000      12.000000   
25%         0.000000       0.000000       1.000000      24.000000   
50%         0.000000       0.000000       1.000000      27.000000   
75%         1.000000       1.000000       1.000000      31.000000   
max         1.000000       1.000000       1.000000      98.000000   

              Smoker         Stroke  HeartDiseaseorAttack   PhysActivity  \
count  202736.000000  202736.000000         202736.000000  202736.000000   
mean        0.442970       0.040886              0.094487       0.756062   
std         0.496738       0.198026              0.292506       0.429457   
min         0.000000       0.000000              0.000000       0.000000   
25%         0.000000       0.000000              0.000000       1.000000   
50%         0.000000       0.000000              0.000000       1.000000   
75%         1.000000       0.000000              0.000000       1.000000   
max         1.000000       1.000000              1.000000       1.000000   

              Fruits        Veggies  HvyAlcoholConsump  AnyHealthcare  \
count  202736.000000  202736.000000      202736.000000  202736.000000   
mean        0.635077       0.811341           0.055969       0.951494   
std         0.481410       0.391239           0.229863       0.214834   
min         0.000000       0.000000           0.000000       0.000000   
25%         0.000000       1.000000           0.000000       1.000000   
50%         1.000000       1.000000           0.000000       1.000000   
75%         1.000000       1.000000           0.000000       1.000000   
max         1.000000       1.000000           1.000000       1.000000   

         NoDocbcCost        GenHlth       MentHlth       PhysHlth  \
count  202736.000000  202736.000000  202736.000000  202736.000000   
mean        0.083779       2.515163       3.200251       4.260782   
std         0.277057       1.070198       7.430349       8.740095   
min         0.000000       1.000000       0.000000       0.000000   
25%         0.000000       2.000000       0.000000       0.000000   
50%         0.000000       2.000000       0.000000       0.000000   
75%         0.000000       3.000000       2.000000       3.000000   
max         1.000000       5.000000      30.000000      30.000000   

            DiffWalk            Sex            Age      Education  \
count  202736.000000  202736.000000  202736.000000  202736.000000   
mean        0.168880       0.439577       8.038513       5.049384   
std         0.374647       0.496337       3.052012       0.985668   
min         0.000000       0.000000       1.000000       1.000000   
25%         0.000000       0.000000       6.000000       4.000000   
50%         0.000000       0.000000       8.000000       5.000000   
75%         0.000000       1.000000      10.000000       6.000000   
max         1.000000       1.000000      13.000000       6.000000   

              Income   ExtraMedTest  ExtraAlcoholTest         Status  
count  202736.000000  202736.000000     202736.000000  202736.000000  
mean        6.053799      -7.416438         -7.566027       0.175751  
std         2.070335      76.031037         75.964939       0.380609  
min         1.000000    -199.000000       -199.000000       0.000000  
25%         5.000000     -55.000000        -55.000000       0.000000  
50%         7.000000       0.000000          0.000000       0.000000  
75%         8.000000      40.000000         40.000000       0.000000  
max         8.000000     199.000000        199.000000       1.000000

In [522]:
# Display the first few rows of the training dataset
styler.draw_box("First 5 rows of the training dataset")

df_train.head()

┌────────────────────────────────────────┐
│  First 5 rows of the training dataset  │
└────────────────────────────────────────┘


HighBP  HighChol  CholCheck  BMI  Smoker  Stroke  HeartDiseaseorAttack  \
0       0         0          1   24       1       0                     0   
1       0         0          1   28       0       0                     0   
2       0         0          1   36       1       0                     0   
3       0         1          1   35       0       0                     0   
4       0         1          1   27       0       0                     0   

   PhysActivity  Fruits  Veggies  HvyAlcoholConsump  AnyHealthcare  \
0             1       0        1                  0              1   
1             1       1        1                  0              1   
2             1       1        0                  0              1   
3             1       1        1                  0              1   
4             1       0        1                  0              1   

   NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex  Age  Education  \
0            0        2         0         0         0    0    8          4   
1            0        1         1         0         0    0    2          6   
2            1        3        30        30         1    0    3          2   
3            0        3         0         0         0    0    8          6   
4            0        3         0         0         0    0    9          5   

   Income  ExtraMedTest  ExtraAlcoholTest  Status  
0       5            60                 0       0  
1       8             0               -64       0  
2       1           -46                 0       0  
3       8           -83              -188       0  
4       4           -58                 0       0

#### 4.3. Data Validation

##### 4.3.1: Handling Missing Values

In [523]:
# Check for missing / numpy.NaN / None values in the training dataset
styler.draw_box("Checking for invalid values in the training dataset")

text = neko.is_nan(df_train)
print(f"{bullet} {text}")

┌───────────────────────────────────────────────────────┐
│  Checking for invalid values in the training dataset  │
└───────────────────────────────────────────────────────┘
>>> The DataFrame does not contain any NaN values.


##### 4.3.2: Detect Noisy Data

In [524]:
zero_one_columns_error = []  # 0-1 columns with errors
range_columns_error = []  # Range columns with errors

# Columhns with 0 and 1 values
zero_one_columns = [
    "Status",
    "HighBP",
    "HighChol",
    "CholCheck",
    "Smoker",
    "Stroke",
    "HeartDiseaseorAttack",
    "PhysActivity",
    "Fruits",
    "Veggies",
    "HvyAlcoholConsump",
    "AnyHealthcare",
    "NoDocbcCost",
    "DiffWalk",
    "Sex",
]

# Columns with range defined in the dataset description
range_column = {
    "Age": (0, 13),
    "Education": (1, 6),
    "Income": (1, 8),
    "BMI": (1, 45),  # 1 to 45 (Explained in the next cell)
    "ExtraMedTest": (-100, 100),
    "ExtraAlcoholTest": (-100, 100),
    "MentHlth": (1, 30),
    "PhysHlth": (1, 30),
    "GenHlth": (1, 5),
}

In [525]:
styler.draw_box("Checking for invalid values for columns with 0 and 1 values")

# Check for invalid values in the training dataset
for column in zero_one_columns:
    # Check for invalid values
    payload = neko.is_data_valid(data=df_train, column=column, valid_values=[0, 1])

    # Append to the error list
    if not payload["is_valid"]:
        zero_one_columns_error.append(column)

    # Show the results
    styled_text = styler.style(payload["message"], bold=True, italic=True)
    print(f"{bullet} {column}: {styled_text}")

┌───────────────────────────────────────────────────────────────┐
│  Checking for invalid values for columns with 0 and 1 values  │
└───────────────────────────────────────────────────────────────┘
>>> Status: All values are within the range.
>>> HighBP: All values are within the range.
>>> HighChol: All values are within the range.
>>> CholCheck: All values are within the range.
>>> Smoker: All values are within the range.
>>> Stroke: All values are within the range.
>>> HeartDiseaseorAttack: All values are within the range.
>>> PhysActivity: All values are within the range.
>>> Fruits: All values are within the range.
>>> Veggies: All values are within the range.
>>> HvyAlcoholConsump: All values are within the range.
>>> AnyHealthcare: All values are within the range.
>>> NoDocbcCost: All values are within the range.
>>> DiffWalk: All values are within the range.
>>> Sex: All values are within the range.


In [526]:
styler.draw_box("Checking for invalid values for columns with the defined range")

# Check each column against its range
for column, (min_value, max_value) in range_column.items():
    # Check for invalid values
    payload = neko.is_data_in_range(df_train, column, min_value, max_value)

    # Append to the error list
    if not payload["is_valid"]:
        range_columns_error.append(column)

    # Show the results
    styled_text = styler.style(payload["message"], bold=True, italic=True)
    print(f"{bullet} {column}: {styled_text}")

┌──────────────────────────────────────────────────────────────────┐
│  Checking for invalid values for columns with the defined range  │
└──────────────────────────────────────────────────────────────────┘
>>> Age: All values are within the range.
>>> Education: All values are within the range.
>>> Income: All values are within the range.
>>> BMI: 3854 invalid values found.
>>> ExtraMedTest: 29515 invalid values found.
>>> ExtraAlcoholTest: 29464 invalid values found.
>>> MentHlth: 140140 invalid values found.
>>> PhysHlth: 127785 invalid values found.
>>> GenHlth: All values are within the range.


**Key Observations**  
The previous data exploration revealed the following key findings:

- Missing Values: The dataset exhibits no missing values. All values in the specified columns (`zero_one_columns`) are confirmed to be either `0` or `1`.
- Outliers: Several columns, including `BMI`, `ExtraMedTest`, `ExtraAlcoholTest`, `MenHlth`, and `PhysHlth`, contain outliers that fall outside the allowed range defined in the assignment instructions.

**Addressing Outliers for Improved Data Quality:**  
- Outliers can potentially influence the results of our analysis. To ensure data quality and integrity, we will employ a common technique of replacing outliers with the mean value of their respective columns. 
- This approach aims to mitigate the impact of extreme values while preserving the overall distribution of the data.


##### 4.3.3: Handling invalid data

In [527]:
styler.draw_box("Handling out of range values for 0-1 columns")

try:
    for column in zero_one_columns_error:
        mean = df_train[column].mean()  # Mean
        df_train[column] = df_train[column].apply(
            lambda x: (0 if x != 0 and x != 1 else x)
        )

        text = f"Column '{column}' has been handled successfully"
        styled_text = styler.style(text, bold=True, italic=True)
        print(f"{bullet} {styled_text}")

    text = "All columns have been handled successfully"
    styled_text = styler.style(text, bold=True, italic=True)
    print(f"{bullet} {styled_text}")
except KeyError as e:
    styler.draw_box(f"Error: Column '{e}' not found in the DataFrame.")

┌────────────────────────────────────────────────┐
│  Handling out of range values for 0-1 columns  │
└────────────────────────────────────────────────┘
>>> All columns have been handled successfully


In [528]:
styler.draw_box("Handling out of range values for ranged columns")

try:
    for column in range_columns_error:
        mean = df_train[column].mean()  # Mean
        df_train[column] = df_train[column].apply(
            lambda x: (
                mean
                if x < range_column[column][0] or x > range_column[column][1]
                else x
            )
        )

        text = f"Column '{column}' has been handled successfully"
        styled_text = styler.style(text, bold=True, italic=True)
        print(f"{bullet} {styled_text}")

    text = "All columns have been handled successfully"
    styled_text = styler.style(text, bold=True, italic=True)
    print(f"{bullet} {styled_text}")
except KeyError as e:
    styler.draw_box(f"Error: Column '{e}' not found in the DataFrame.")

┌───────────────────────────────────────────────────┐
│  Handling out of range values for ranged columns  │
└───────────────────────────────────────────────────┘
>>> Column 'BMI' has been handled successfully
>>> Column 'ExtraMedTest' has been handled successfully
>>> Column 'ExtraAlcoholTest' has been handled successfully
>>> Column 'MentHlth' has been handled successfully
>>> Column 'PhysHlth' has been handled successfully
>>> All columns have been handled successfully


##### 4.3.4: Data Revalidation

In [536]:
# Display the status of invalid values after handling

styler.draw_box(
    "Checking for invalid values for columns with the defined range after handling"
)

# Check each column against its range
for column, (min_value, max_value) in range_column.items():
    # Check for invalid values
    payload = neko.is_data_in_range(df_train, column, min_value, max_value)

    # Append to the error list
    if not payload["is_valid"]:
        range_columns_error.append(column)

    # Show the results
    styled_text = styler.style(payload["message"], bold=True, italic=True)
    print(f"{bullet} {column}: {styled_text}")

# Spacing
print("")

styler.draw_box("Handling out of range values for 0-1 columns after handling")

try:
    for column in zero_one_columns_error:
        mean = df_train[column].mean()  # Mean
        df_train[column] = df_train[column].apply(
            lambda x: (0 if x != 0 and x != 1 else x)
        )

        text = f"Column '{column}' has been handled successfully"
        styled_text = styler.style(text, bold=True, italic=True)
        print(f"{bullet} {styled_text}")

    text = "All columns have been handled successfully"
    styled_text = styler.style(text, bold=True, italic=True)
    print(f"{bullet} {styled_text}")
except KeyError as e:
    styler.draw_box(f"Error: Column '{e}' not found in the DataFrame.")

┌─────────────────────────────────────────────────────────────────────────────────┐
│  Checking for invalid values for columns with the defined range after handling  │
└─────────────────────────────────────────────────────────────────────────────────┘
>>> Age: All values are within the range.
>>> Education: All values are within the range.
>>> Income: All values are within the range.
>>> BMI: All values are within the range.
>>> ExtraMedTest: All values are within the range.
>>> ExtraAlcoholTest: All values are within the range.
>>> MentHlth: All values are within the range.
>>> PhysHlth: All values are within the range.
>>> GenHlth: All values are within the range.

┌───────────────────────────────────────────────────────────────┐
│  Handling out of range values for 0-1 columns after handling  │
└───────────────────────────────────────────────────────────────┘
>>> All columns have been handled successfully


In [534]:
# Display the overview of the training dataset after handling invalid values
df_train.describe()

HighBP       HighChol      CholCheck            BMI  \
count  202736.000000  202736.000000  202736.000000  202736.000000   
mean        0.429060       0.424631       0.962621      27.892295   
std         0.494943       0.494288       0.189688       5.337065   
min         0.000000       0.000000       0.000000      12.000000   
25%         0.000000       0.000000       1.000000      24.000000   
50%         0.000000       0.000000       1.000000      27.000000   
75%         1.000000       1.000000       1.000000      31.000000   
max         1.000000       1.000000       1.000000      45.000000   

              Smoker         Stroke  HeartDiseaseorAttack   PhysActivity  \
count  202736.000000  202736.000000         202736.000000  202736.000000   
mean        0.442970       0.040886              0.094487       0.756062   
std         0.496738       0.198026              0.292506       0.429457   
min         0.000000       0.000000              0.000000       0.000000   
25%         0.000000       0.000000              0.000000       1.000000   
50%         0.000000       0.000000              0.000000       1.000000   
75%         1.000000       0.000000              0.000000       1.000000   
max         1.000000       1.000000              1.000000       1.000000   

              Fruits        Veggies  HvyAlcoholConsump  AnyHealthcare  \
count  202736.000000  202736.000000      202736.000000  202736.000000   
mean        0.635077       0.811341           0.055969       0.951494   
std         0.481410       0.391239           0.229863       0.214834   
min         0.000000       0.000000           0.000000       0.000000   
25%         0.000000       1.000000           0.000000       1.000000   
50%         1.000000       1.000000           0.000000       1.000000   
75%         1.000000       1.000000           0.000000       1.000000   
max         1.000000       1.000000           1.000000       1.000000   

         NoDocbcCost        GenHlth       MentHlth       PhysHlth  \
count  202736.000000  202736.000000  202736.000000  202736.000000   
mean        0.083779       2.515163       5.412404       6.946364   
std         0.277057       1.070198       6.575482       7.598298   
min         0.000000       1.000000       1.000000       1.000000   
25%         0.000000       2.000000       3.200251       4.260782   
50%         0.000000       2.000000       3.200251       4.260782   
75%         0.000000       3.000000       3.200251       4.260782   
max         1.000000       5.000000      30.000000      30.000000   

            DiffWalk            Sex            Age      Education  \
count  202736.000000  202736.000000  202736.000000  202736.000000   
mean        0.168880       0.439577       8.038513       5.049384   
std         0.374647       0.496337       3.052012       0.985668   
min         0.000000       0.000000       1.000000       1.000000   
25%         0.000000       0.000000       6.000000       4.000000   
50%         0.000000       0.000000       8.000000       5.000000   
75%         0.000000       1.000000      10.000000       6.000000   
max         1.000000       1.000000      13.000000       6.000000   

              Income   ExtraMedTest  ExtraAlcoholTest         Status  
count  202736.000000  202736.000000     202736.000000  202736.000000  
mean        6.053799       0.388297          0.251700       0.175751  
std         2.070335      49.452401         49.476287       0.380609  
min         1.000000    -100.000000       -100.000000       0.000000  
25%         5.000000     -24.000000        -25.000000       0.000000  
50%         7.000000       0.000000          0.000000       0.000000  
75%         8.000000      24.000000         24.000000       0.000000  
max         8.000000     100.000000        100.000000       1.000000

<center><em><sub>─────── End Of Section ───────</sub></em></center>


In [529]:
# # Step 5: Feature scaling

# # Min-max scaling is applied to maintain uniformity within the data range.
# # Formular: X = (X - min) / (max - min)
# min_max_scaler = MinMaxScaler()

# columns_to_scale = ["BMI", "ExtraMedTest", "ExtraAlcoholTest", "MentHlth", "PhysHlth"]
# df[columns_to_scale] = min_max_scaler.fit_transform(df[columns_to_scale])

# # Inspect data after scaling
# df.describe().round(3)

In [530]:
# # Step 6: Handle Outliers

# # Grid initilization
# rows, cols = len(df.columns) // 4, 4

# fig, axes = plt.subplots(rows, cols, figsize=(12, rows * 3))

# for i, (column, ax) in enumerate(zip(df.columns, axes.flatten())):
#     sns.kdeplot(data=df[column], ax=ax)
#     ax.set_title(f"KDE Plot of {column}")
#     ax.set_xlabel(column)

# plt.tight_layout()
# plt.show()

In [531]:
# # Step 7: Handle categorical data

# # Create one-hot encoding
# age_dummies = pd.get_dummies(df["Age"]).astype(int)
# gen_health_dummies = pd.get_dummies(df["GenHlth"]).astype(int)
# education_dummies = pd.get_dummies(df["Education"]).astype(int)
# income_dummies = pd.get_dummies(df["Income"]).astype(int)

# # Rename the columns
# gen_health_dummies.columns = [
#     "Genhlth_1",
#     "Genhlth_2",
#     "Genhlth_3",
#     "Genhlth_4",
#     "Genhlth_5",
# ]

# age_dummies.columns = [
#     "Age_18_24",
#     "Age_25_29",
#     "Age_30_34",
#     "Age_35_39",
#     "Age_40_44",
#     "Age_45_49",
#     "Age_50_54",
#     "Age_55_59",
#     "Age_60_64",
#     "Age_65_69",
#     "Age_70_74",
#     "Age_75_79",
#     "Age_80_or_older",
# ]

# education_dummies.columns = [
#     "Educ_Never",
#     "Educ_G1_8",
#     "Educ_G9_11",
#     "Educ_G12_GED",
#     "Educ_Col_1_3",
#     "Educ_Col_4_more",
# ]

# income_dummies.columns = [
#     "Inc_1",
#     "Inc_2",
#     "Inc_3",
#     "Inc_4",
#     "Inc_5",
#     "Inc_6",
#     "Inc_7",
#     "Inc_8",
# ]

# # Drop the original
# df.drop(["Education", "Income", "Age", "GenHlth"], axis=1, inplace=True)

# # Concatenate the original DataFrame
# df = pd.concat(
#     [df, age_dummies, gen_health_dummies, education_dummies, income_dummies], axis=1
# )

In [532]:
# df.to_csv("../../data/processed/data_train_processed.csv")